embulk-util-throttle-helper/src/main/java/org/embulk/util/throttle_helper/QuotaRateLimiter.java

```diff
@@ -0,0 +1,157 @@
+ *
+ * @author instcode
+ * @author thangnc
+ */
+@ThreadSafe
+public class QuotaRateLimiter
+{
+    /**
+     * This is to prevent multiple threads from updating {@link #remainingLimit} &
+     * {@link #lastRefreshTime} wrongly
+     */
+    private Lock lock = new ReentrantLock();
+
+    /**
+     * The number of permits per the given time window
+     */
+    private int rateLimit;
+
+    /**
+     * The amount of time that affects the rate limit, i.e. the number of
+     * operations should not exceed the rate limit within this time window.
+     * This is also the amount of time that the rate limit gets refreshed.
+     */
+    private int timeWindowInMillis;
+
+    /**
+     * The number of available permits to give. If it is a positive number, it's
+     * the number of remaining permits for the current time window. Otherwise, it
+     * indicates that there are pending reserved permits that may affect the time to
+     * wait of incoming requests. For example, assume that the rate limit is 5 in
+     * 10 seconds as starting point, then if there are 15 requests come at once,
+     * there will be at least 5 requesters should have to wait for 20 seconds and
+     * the remaining limit is -5
+     */
+    private volatile int remainingLimit;
+
+    /**
+     * The last rate limit refresh time (ms since epoch)
+     */
+    private volatile long lastRefreshTime;
+
+    /**
+     * Instantiates a new Time window rate limiter.
+     *
+     * @param rateLimit           the rate limit
+     * @param timeWindowInSeconds the time window in seconds
+     */
+    public QuotaRateLimiter(int rateLimit, int timeWindowInSeconds)
+    {
+        this(rateLimit, rateLimit, timeWindowInSeconds, System.currentTimeMillis());
+    }
+
+    /**
+     * Instantiates a new Time window rate limiter.
+     *
+     * @param rateLimit           the rate limit
+     * @param remainingLimit      the remaining limit
+     * @param timeWindowInSeconds the time window in seconds
+     * @param lastRefreshTime     the last refresh time
+     */
+    public QuotaRateLimiter(int rateLimit, int remainingLimit, int timeWindowInSeconds, long lastRefreshTime)
+    {
+        this.rateLimit = rateLimit;
+        this.remainingLimit = remainingLimit;
+        this.timeWindowInMillis = timeWindowInSeconds * 1000;
+        this.lastRefreshTime = lastRefreshTime;
+    }
+
+    /**
+     * Request a number of permits and may be blocking until the request is granted.
+     *
+     * @param permits The permits
+     * @return The time spent sleeping in milliseconds
+     */
+    public long acquire(final int permits)
+    {
+        long millisToSleepTime = reserveAndGetWaitTime(permits);
+        Uninterruptibles.sleepUninterruptibly(millisToSleepTime, TimeUnit.MILLISECONDS);
+        return millisToSleepTime;
+    }
+
+    /**
+     * Gets remaining limit
+     *
+     * @return the remaining limit
+     */
+    @VisibleForTesting
+    public int getRemainingLimit()
```

Comment:
1. If this is to be called from plugin's main code, I'd suggest to name the method or at least comment (Javadoc) on it to describe the "remaining limit" is per the current time window.
2. If this is not to be called from plugin's main code (really only for testing), I'd suggest to name it like that (e.g. `getRemainingLimitForTesting`, not (only) `@VisibleForTesting`, to discourage developers not to call this from their non-testing code.

Reply:
I see. Changed to `getRemainingLimitForTesting`.